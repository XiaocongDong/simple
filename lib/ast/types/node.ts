export enum NODE_TYPE {
  TOKEN = 'TOKEN',
  IDENTIFIER = 'IDENTIFIER',
  BOOLEAN_LITERAL = 'BOOLEAN_LITERAL',
  SINGLE_EXPRESSION = 'SINGLE_EXPRESSION',
  VARIABLE_MODIFIER = 'VARIABLE_MODIFIER',
  VARIABLE_DECLARATOR = 'VARIABLE_DECLARATOR',
  VARIABLE_STATEMENT = 'VARIABLE_STATEMENT',
}
