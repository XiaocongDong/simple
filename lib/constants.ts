export const SPACE = ' '
export const NEW_LINE = '\n'

// keywords
export const LET = 'let'
export const CONST = 'const'
export const BREAK = 'break'
export const CONTINUE = 'continue'
export const IF = 'if'
export const ELSE = 'else'
export const WHILE = 'while'
export const FUNCTION = 'function'
export const UNDEFINED = 'undefined'
export const NULL = 'null'
export const TRUE = 'true'
export const FALSE = 'false'
export const NEW = 'new'
export const CLASS = 'class'
