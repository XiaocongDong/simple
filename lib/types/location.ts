export interface ILocation {
  column: number
  line: number
}

export interface ILocationRange {
  start: ILocation
  end: ILocation
}
