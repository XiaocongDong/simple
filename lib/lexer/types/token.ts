import { ILocationRange } from "./location";

export enum TOKEN_TYPE {
  IDENTIFER = 'IDENTIFIER',

  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',

  // keywords
  LET = 'LET',
  CONST = 'CONST',
  BREAK = 'BREAK',
  CONTINUE = 'CONTINUE',
  IF = 'IF',
  ELSE = 'ELSE',
  WHILE = 'WHILE',
  FUNCTION = 'FUNCTION',
  UNDEFINED = 'UNDEFINED',
  NULL = 'NULL',
  TRUE = 'TRUE',
  FALSE = 'FALSE',
  NEW = 'NEW',

  LEFT_PAREN = 'LEFT_PAREN',
  RIGHT_PAREN = 'RIGHT_PAREN',
  SEMI_COLON = 'SEMICOLON',
  LEFT_CURLY_BRACE = 'LEFT_CURLY_BRACE',
  RIGHT_CURLY_BRACE = 'RIGHT_CURLY_BRACE',
  LEFT_SQUARE_BRACKET = 'LEFT_SQUARE_BRACKET',
  RIGHT_SQUARE_BRACKET = 'RIGHT_SQUARE_BRACKET',
  COMMA = 'COMMA',
  DOT = 'DOT',
  COLON = 'COLON',

  // mathematical operators
  PLUS = 'PLUS',
  PLUS_PLUS = 'PLUS_PLUS',
  MINUS = 'MINUS',
  MINUS_MINUS = 'MINUS_MINUS',
  MULTIPLY = 'MULTIPLY',
  DIVIDE = 'DIVIDE',
  PLUS_ASSIGN = 'PLUS_ASSIGN',
  MINUS_ASSIGN = 'MINUS_ASSIGN',
  MULTIPLY_ASSIGN = 'MULTIPLY_ASSIGN',
  DIVIDE_ASSIGN = 'DIVIDE_ASSIGN',
  ASSIGN = 'ASSIGN',
  LESS_THAN = 'LESS_THAN',
  LESS_EQUAL_THAN = 'LESS_EQUAL_THAN',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_EQUAL_THAN = 'GREATER_EQUAL_THEN',

  // logical operators
  OR = 'OR',
  AND = 'AND',
  EQUAL = "EQUAL"
}

export interface IToken {
  type: TOKEN_TYPE
  value: string
  range: ILocationRange
}
