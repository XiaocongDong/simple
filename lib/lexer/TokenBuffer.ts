import { IToken } from "./types/token"

class TokenBuffer {
  private tokens: Array<IToken> = []
  private cursor: number = 0

  peek() {
    return this.tokens[this.cursor]
  }

  read() {
    return this.tokens[this.cursor++]
  }

  write(token: IToken) {
    this.tokens.push(token)
  }

  getCursor() {
    return this.cursor
  }

  setCursor(cursor: number) {
    this.cursor = cursor
  }

  toJSON(): Array<IToken> {
    return this.tokens
  }

  isEmpty(): boolean {
    return this.cursor === this.tokens.length
  }
}

export default TokenBuffer
