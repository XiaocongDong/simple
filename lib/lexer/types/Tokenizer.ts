import { TOKEN_TYPE } from "./token"

export type IState = string
export type ITokenTypeGenerator = (value: string) => TOKEN_TYPE
export type IStateConfig = {
  isEnd: boolean
  TokenType?: TOKEN_TYPE|ITokenTypeGenerator
  transitions?: Array<ITransition>
}
export type IStatesConfig = Record<IState, IStateConfig>

interface ITransition {
  checker: RegExp|string|((ch: string) => boolean)
  state: string
}

export type IConfig = {
  states: IStatesConfig
  initialState: IState
}
