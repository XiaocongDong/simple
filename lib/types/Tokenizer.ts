import { TOKEN_TYPE } from "./token"

export type IState = string
export type ITokenTypeGenerator = (value: string) => TOKEN_TYPE
export type IStateConfig = {
  isEnd: boolean
  TokenType?: TOKEN_TYPE|ITokenTypeGenerator
  transitions?: Array<ITransition>
}

interface ITransition {
  checker: RegExp|string|((ch: string) => boolean)
  state: string
}

export type IConfig = {
  states: Record<IState, IStateConfig>
  initialState: IState
}
