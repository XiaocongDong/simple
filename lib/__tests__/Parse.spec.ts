import { IToken, TOKEN_TYPE } from "../lexer/types/token"
import { LET } from "../constants"
import TokenBuffer from "../lexer/TokenBuffer"
import parse from "../parser"

const testCases: Array<{tokens: Array<IToken>, description: string}> = [
  {
    tokens: [
      {
        type: TOKEN_TYPE.LET,
        value: LET,
        range: {
          start: {
            line: 0,
            column: 1
          },
          end: {
            line: 0,
            column: 3
          }
        }
      },
      {
        type: TOKEN_TYPE.IDENTIFER,
        value: 'a',
        range: {
          start: {
            line: 0,
            column: 5
          },
          end: {
            line: 0,
            column: 5
          }
        }
      },
      {
        type: TOKEN_TYPE.EQUAL,
        value: '=',
        range: {
          start: {
            line: 0,
            column: 7
          },
          end: {
            line: 0,
            column: 7
          }
        }
      },
      {
        type: TOKEN_TYPE.NUMBER_LITERAL,
        value: '10',
        range: {
          start: {
            line: 0,
            column: 9
          },
          end: {
            line: 0,
            column: 10
          }
        }
      }
    ],
    description: 'variable declaration'
  }
]

let tokenBuffer = null
beforeEach(() => {
  tokenBuffer = new TokenBuffer() as any
})

describe('parse', () => {
  testCases.forEach(({ tokens, description }) => {
    it(description, () => {
      tokenBuffer.tokens = tokens
      const ast = parse(tokenBuffer)
      expect(ast).toMatchSnapshot()
    })
  })
})
