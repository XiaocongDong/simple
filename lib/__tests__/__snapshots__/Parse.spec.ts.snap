// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse variable declaration 1`] = `
Node {
  "children": Array [],
  "loc": Object {
    "end": null,
    "start": null,
  },
}
`;
